### DISCUSSION TITLE
periodic boundary condition implementation

### DISCUSSION THREAD

**Discussion by karthichockalingam** [idaholab/moose#24984] [Q&A General] [answered]
Hi,

This a general question to understand how periodic boundary conditions are implemented in MOOSE.

My initial thought was to come up with a relation between the periodic nodes (so I can eliminate one set) on a regular rectilinear mesh. 
 
Let the relation between the nodes be as follows,
x = P x’
 
Say for 1-D problem with two elements
(1)-------------(2)------------(3)
 
P = [1 0, 0 1, 1 0]
x = [u1 u2 u3]
x’ = [u1 u2]
 
and solve
[P^T A P] x’ = P^T b
 
I don’t think [P^T A P] is deterministic. So I am missing something?

Thank you very much for your help.

Kind regards,
Karthik.

**Comment by lindsayad** [comment]
@roystgnr

**Comment by roystgnr** [comment]
`[P^T A P]` is indeed how we do things under the hood, in libMesh.  The final solution is then well-defined and completely determined (or at least as close to that as you can get in a world where threading and MPI have to assume associativity for speed but yet floating-point operations are not in general associative).  But you're correct that the specific `P` isn't yet well-defined, because e.g. for any pair of Degrees of Freedom with a 1-to-1 mapping you can always pick either one of them to be the "independent" DoF and the other to be the "constrained" DoF.  It's an especially giant pain when you have multiple periodic constraints intersecting, like at the corners of a periodic square or the corners or edges of a periodic cube.  On a distributed mesh a node's owner might not even see the node that (via a constraint of a constraint) it is ultimately tied to.

On adaptively refined grids, we always constrain finer elements' DoFs to coarser ones, so we can do adaptive constraints and periodic constraints in one swoop ... but when that's not enough to disambiguate things we just use element and node `id()` comparisons to make the remaining choices.  This is deterministic in the critical "no disagreements between different processors in parallel" sense, and technically it's deterministic in the sense that we never assign ids randomly, but since ids can end up being influenced by input file, mesh partitioning, mesh renumbering settings, etc, it's not very "deterministic" in the broad spirit of that term.

**Reply by karthichockalingam** [reply]
@roystgnr Thank you for your detailed response. Yes, ```P``` is not well defined as you pointed out in the case of corners of a periodic square. 

However, I have a fundamental question. Please refer to the below-attached image. After applying periodic boundary conditions on a 1-D domain the final resulting matrix (```P^T A P```) in equation (3.14) has a zero determinant. Why is that?


![Screenshot 2023-07-18 at 18 54 58](https://github.com/idaholab/moose/assets/39724615/478fa5b0-1e25-4105-a43f-039b5e7dc364)

**Reply by roystgnr** [reply, marked as answer]
For that differential equation, the infinite-dimensional linear operator *itself* has a null space!  `-Δu = 0` is true for any `û≡C` with constant C, and this constant satisfies the periodic boundary conditions too.  So for any solution `u` there's a whole family of solutions `u+cû` and the problem isn't able to distinguish between them.  You need something to further constrain the problem (make it a time-dependent PDE with initial conditions, for example) if you want a well-defined unique solution, or if you actually don't care which of the family of solutions you come up with then you can add an artificial constraint or use a null-space-aware algebraic solver.

Under the hood in libMesh, there's one less interesting catch: we use a square (non-invertible) rather than a rectangular `P`, so we end up with a `P^T A P` that has zero-rows and -columns for the constrained DoFs, rather than a smaller matrix of only unconstrained DoFs' terms.  We have various options (e.g. throwing the constraint equations themselves into those rows) for how to solve that, and we hope to make "just use the smaller matrix" an option too in the future.

**Reply by karthichockalingam** [reply]
Thank you so much for your response. Now it all makes sense!
